WcaOnRails/app/models/result.rb

```diff
@@ -1,15 +1,86 @@
 # frozen_string_literal: true
 class Result < ActiveRecord::Base
+  include ResultMethods
+
   self.table_name = "Results"
 
   belongs_to :competition, foreign_key: :competitionId
+  belongs_to :country, foreign_key: :countryId
   belongs_to :person, -> { current }, primary_key: :wca_id, foreign_key: :personId
   belongs_to :round, foreign_key: :roundId
   belongs_to :event, foreign_key: :eventId
+  belongs_to :format, foreign_key: :formatId
 
   scope :podium, -> { joins(:round).merge(Round.final_rounds).where(pos: [1..3]).where("best > 0") }
   scope :winners, -> { joins(:round, :event).merge(Round.final_rounds).where("pos = 1 and best > 0").order("Events.rank") }
 
+  validate :best_and_average
+  def best_and_average
+    return errors.add(:competitionId, "invalid") unless competition
+    return errors.add(:countryId, "invalid") unless country
+    return errors.add(:roundId, "invalid") unless round
+    return errors.add(:eventId, "invalid") unless event
+    return errors.add(:formatId, "invalid") unless format
+
+    solve_times.each_with_index do |solve_time, i|
+      unless solve_time.valid?
+        errors.add(:"value#{i + 1}", solve_time.errors.messages[:base].join(" "))
+      end
+    end
+
+    if solve_times.all?(&:skipped?)
+      errors.add(:base, "Cannot skip all solves.")
+      return
```

Comment:
Some issues arise if you try to keep checking for errors:

- We need a valid `roundId` and `formatId` to know how many solves we expect the Result to have. Without those, we could check the `best`, but we cannot check the `average`.
- If a round only had 3 solves, and there are 4 solves entered, I don't know what it would mean for us to check the `average`.
- If a round had 5 solves, but some of them were skipped, I don't know what it would mean for us to check the `average`.
- We need to know the `eventId` to compute `average`, because 333fm's average field is so funky.

Basically, there are some preconditions we need to satisfy before it makes sense to talk about the result's "average". Maybe there's a better way of thinking of this, let me know what you think.

3We could rewrite this to always check `best`.